pipeline_server/src/runner.rs

```diff
@@ -0,0 +1,479 @@
+                    return Err(AnyError::msg(mtch.as_str().to_string()));
+                };
+            }
+
+            if start.elapsed() > STARTUP_TIMEOUT {
+                let log = Self::log_suffix(log_file_path).await;
+                return Err(AnyError::msg(format!("waiting for pipeline initialization status timed out after {STARTUP_TIMEOUT:?}\n{log}")));
+            }
+            sleep(Duration::from_millis(100)).await;
+        }
+    }
+
+    /// Create Prometheus config file for a pipeline.
+    async fn create_prometheus_config(
+        &self,
+        project_name: &str,
+        project_id: ProjectId,
+        pipeline_id: PipelineId,
+        port: u16,
+    ) -> AnyResult<()> {
+        let config = format!(
+            r#"- targets: [ "localhost:{port}" ]
+  labels:
+    project_name: "{project_name}"
+    pipeline_id: {pipeline_id}
+    project_id: {project_id}"#
+        );
+        fs::write(
+            self.config.prometheus_pipeline_config_file(pipeline_id),
+            config,
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /*
+    async fn create_topics(&self, config_yaml: &str) -> AnyResult<(KafkaResources, String)> {
+        let mut config: ControllerConfig = serde_yaml::from_str(config_yaml)
+            .map_err(|e| AnyError::msg(format!("error parsing pipeline configuration: {e}")))?;
+
+        for (_, input_config) in config.inputs.iter_mut() {
+            if input_config.transport.name == "kafka" {
+                if let Some(topics) = input_config.transport.config.get("topics") {
+                    let topics_string = serde_yaml::to_string(topics);
+                    let topics: Vec<String> = serde_yaml::from_value(topics)
+                        .map_err(|e| AnyError::msg(format!("error parsing Kafka topic list '{topics_string}': {e}")))?;
+                    for topic in topics.iter_mut() {
+                        if topic == "?" {
+                            *topic = self.generate_topic_name("input_");
+                        }
+                    }
+                    input_config.transport.config.set("topics", serde_yaml::to_value(topics))
+                }
+            }
+        }
+
+        for (_, output_config) in config.outputs.iter_mut() {
+            if input_config.transport.name == "kafka" {
+                if let Some(YamlValue::String(topic) = input_config.transport.config.get_mut("topic") {
+                    if topic == "?" {
+                        topic = Self::generate_topic_name("output_");
+                    }
+                }
+            }
+        }
+
+        let new_config = serde_yaml::to_string(config);
+
+        Ok((kafka_resources, new_config))
+
+    }
+    */
+
+    async fn log_suffix_inner(log_file_path: &Path) -> AnyResult<String> {
+        let mut buf = Vec::with_capacity(10000);
+
+        let mut file = File::open(log_file_path).await?;
+
+        Pin::new(&mut file).start_seek(SeekFrom::End(-LOG_SUFFIX_LEN))?;
```

Comment:
in lines?

Reply:
Bytes actually, which is a good catch, because I myself got confused and wrote "lines" in the comment few lines below.

It would be better to seek N lines from the end, but I can't think of a simple way to do this.